cpp/tests/shared_fcntl_mutex_test.cc

```diff
@@ -0,0 +1,516 @@
+    return std::move(f)();
+  };
+
+  // Run the message loop until exit (or error)
+  for (auto loop = true; loop;) {
+    ARROW_ASSIGN_OR_RAISE(auto msg, pipe_comms.receive(context));
+    ARROW_ASSIGN_OR_RAISE(loop, MaybeRunInThread(msg, [&]() { return HandleMsg(msg); }));
+  }
+
+  return Status::OK();
+}
+
+// Spawns some children running child_loop
+// and returns control to the parent process
+template <std::size_t CHILDREN>
+arrow::Status SpawnChildren(std::array<PipeComms, CHILDREN>& comms,
+                            std::array<pid_t, CHILDREN>& child_pid) {
+  for (std::size_t c = 0; c < CHILDREN; ++c) {
+    ARROW_ASSIGN_OR_RAISE(comms[c], PipeComms::Create());
+  }
+
+  auto path = temp_directory_name();
+  if (!fs::create_directory(path))
+    return Status::IOError("Unable to create ", path.native());
+  auto lock_name = path / "table.lock";
+
+  for (std::size_t c = 0; c < CHILDREN; ++c) {
+    if (child_pid[c] = fork(); child_pid[c] == 0) {
+      // child process
+      // run child loop and exit appropriately
+      comms[c].close_unused_ends(PipeComms::CHILD);
+      auto status = [&]() {
+        try {
+          return child_loop(comms[c], lock_name.native());
+        } catch (std::exception& e) {
+          return Status::UnknownError(e.what());
+        }
+      }();
+
+      if (status.ok()) exit(EXIT_SUCCESS);
+      ARROW_LOG(WARNING) << "Child " << c << " failed: " << status;
+      exit(EXIT_FAILURE);
+    } else if (child_pid[c] == -1) {
+      return Status::IOError("Fork of Child ", c, " failed");
+    }
+    // parent process. do pipe end cleanup
+    comms[c].close_unused_ends(PipeComms::PARENT);
+  }
+
+  return Status::OK();
+}
+
+// Issues exit instructions to the child process
+// and waits for it to shutdown
+template <std::size_t CHILDREN>
+arrow::Status ShutdownChildren(std::array<PipeComms, CHILDREN>& comms,
+                               std::array<pid_t, CHILDREN>& child_pid) {
+  for (std::size_t c = 0; c < NCHILDREN; ++c) {
+    int status;
+    ARROW_RETURN_NOT_OK(comms[c].send("exit", PipeComms::PARENT));
+    ARROW_RETURN_NOT_OK(comms[c].expect("ok", PipeComms::PARENT));
+
+    if (auto w_pid = waitpid(child_pid[c], &status, 0); w_pid == -1) {
+      return Status::IOError("Waiting for child ", c, " failed");
+    } else if (status != EXIT_SUCCESS) {
+      return Status::IOError("Child ", c, " failed with exit code", status);
+    }
+  }
+
+  return Status::OK();
+}
+
+static constexpr auto PARENT_CONTEXT = PipeComms::PARENT;
+
+// Test that acquiring read locks increases the number of fcntl readers
+TEST(SharedFcntlMutexTest, FcntlReaders) {
+  std::array<PipeComms, NCHILDREN> comms;
+  std::array<pid_t, NCHILDREN> child_pid;
+
+  ASSERT_OK(SpawnChildren(comms, child_pid));
+  constexpr std::size_t NREADLOCKS = 3;
+
+  // Base case for no locks
+  ASSERT_OK(comms[0].send(kQueryFcntlReaders, PARENT_CONTEXT));
+  ASSERT_OK(comms[0].expect("ok 0", PARENT_CONTEXT));
```

Comment:
Having to pass PARENT_CONTEXT everywhere makes the code somewhat repetitive. It might be nice to restructure things in such a way that the comms object captures just the endpoint that's actually executing. Possibly it could be integrated with spawning the children: have SpawnChildren make the `pipe` calls, and in the parent branch create a PipeComms that captures just the parent's endpoints, and in the child branch create a PipeCommcs that captures the child's endpoints.